test/parallel/test-worker-init-failure.js

```diff
@@ -0,0 +1,63 @@
+'use strict';
+const common = require('../common');
+const assert = require('assert');
+const child_process = require('child_process');
+
+// Test that workers fail with meaningful error message
+// when their initialization fails.
+
+if (common.isWindows) {
+  common.skip('ulimit does not work on Windows.');
+}
+
+// A reasonably low fd count. An empty node process
+// creates around 30 fds for its internal purposes,
+// so making it too low will crash the process early,
+// making it too high will cause too much resource use.
+const OPENFILES = 128;
+
+// Double the open files - so that some workers fail for sure.
+const WORKERCOUNT = 256;
+
+if (process.argv[2] === 'child') {
+  const { Worker } = require('worker_threads');
+  for (let i = 0; i < WORKERCOUNT; ++i) {
+    const worker = new Worker(
+      'require(\'worker_threads\').parentPort.postMessage(2 + 2)',
+      { eval: true });
+    worker.on('message', (result) => {
+      assert.strictEqual(result, 4);
+    });
+    worker.on('error', (e) => {
+      assert.match(e.message, /Worker initialization failure: EMFILE/);
+      assert.strictEqual(e.code, 'ERR_WORKER_INIT_FAILED');
+    });
+  }
+
+} else {
+  // Limit the number of open files, to force workers to fail
+  let testCmd = `ulimit -n ${OPENFILES} && `;
+
+  testCmd += `${process.execPath} ${__filename} child`;
+  const cp = child_process.exec(testCmd);
+
+  let stdout = '';
+  let stderr = '';
+
+  cp.on('exit', common.mustCall((code, signal) => {
+    assert.strictEqual(code, 0);
```

Comment:
This will result in a frustrating error message on failure:

```
=== release test-worker-init-failure ===                   
Path: parallel/test-worker-init-failure
assert.js:102
  throw new AssertionError(obj);
  ^

AssertionError [ERR_ASSERTION]: Expected values to be strictly equal:

1 !== 0

    at ChildProcess.<anonymous> (/Users/trott/io.js/test/parallel/test-worker-init-failure.js:48:12)
    at ChildProcess.<anonymous> (/Users/trott/io.js/test/common/index.js:362:15)
    at ChildProcess.emit (events.js:321:20)
    at Process.ChildProcess._handle.onexit (internal/child_process.js:276:12) {
  generatedMessage: true,
  code: 'ERR_ASSERTION',
  actual: 1,
  expected: 0,
  operator: 'strictEqual'
}
Command: out/Release/node /Users/trott/io.js/test/parallel/test-worker-init-failure.js
```

That tells me that it failed, but not why. To potentially provide a more helpful message, maybe something like this?:

```suggestion
    assert.strictEqual(code, 0, `\nSTDERR: ${stderr}\nSTDOUT: ${stdout}\n`);
```

(Or maybe easier, move [lines 50-53](https://github.com/nodejs/node/pull/31929/files#diff-0ed4e124d7875d8f00dee23b7c9994e3R50-R53) above this assertion?

Reply:
@Trott, I am able to recreate the problem in my `mac` with 8 CPUs. I believe I know the issue now.

I developed this program originally in a linux box with 100+ CPUs. There, almost all threads will start together, so they all fail at worker initialisation phase itself.

In case of 8 CPUs, few threads will start in one batch where as other threads will wait for the started threads to complete.

The started threads could go ahead with the JS execution, and based on the `fd` exhaustion, they can fail at some other point, not in the worker initialisation phase.

That is why we get `EMFILE: too many open files, uv_cwd`

I ran the test 1K times in the failing box, and got only these 2 types of failures.

To fix this, I changed the assertion check as:

```js
     assert.match(e.message, /EMFILE/);
      assert.ok(e.code === 'ERR_WORKER_INIT_FAILED' || e.code === 'EMFILE');
```

I could do more stricter version of:
```js
assert.match(e.message, /Worker initialization failure: EMFILE|EMFILE: too many open files, uv_cwd/);
```

as well, but not sure if there are more possible locations of failure. `/EMFILE/` is generic, and fits the purpose of the test also. Are you fine with this?